proposals/0000-existentials.rst

```diff
@@ -0,0 +1,504 @@
+      in ``pats`` leaks into the type of the expression; this is like a skolem-escape.
+
+      (More technically) When inferring the type of a lambda-expression ``\ pats -> e``: We must bump the
+      type-checker level before checking ``e``. This is because variables bound in ``pats`` might be mentioned
+      in ``Witness``\ es in the type of ``e``. Then, when forming the type of the whole lambda-expression,
+      we promote the inferred type of ``e``; if any variables bound in ``pats`` are encountered, issue an
+      error.
+
+   #. When trying to satisfy a class constraint ``[W] C (exists (a :: k). ty)``, instead solve
+      ``[W] forall (a :: k). C ty``. For multi-parameter type classes, apply this treatment one
+      argument at a time. Note that this treatment applies to ``~``, which is solved like any
+      other type class (recall that lifted ``~`` is distinct from unlifted ``~#``, to which this
+      treatment does not apply).
+
+   #. When trying to satisfy a class constraint ``[W] C (C' /\ ty)``, instead solve
+      ``[W] C' => C ty``. That is, assume ``C'`` as a given and then solve ``[W] C ty``.
+      For multi-parameter type classes, apply this treatment one argument at a time.
+
+#. **Core language.** There are several modifications to the Core language necessary to
+   support this proposal. The notes here echo the design in the paper_, Section 5.
+
+   1. The ``exists`` type would need to be added to Core as a new constructor of ``Type``::
+
+        | ExistsTy TyVar Type
+          -- typing rule:
+          -- Γ ⊢ ki : Type
+          -- Γ, tv:ki ⊢ inner_ty : TYPE rep
+          -- tv # rep
+          -- ------------------------------------
+          -- Γ ⊢ ExistsTy (tv:ki) inner_ty : TYPE rep
+
+   #. The ``Witness`` type would need to be added to Core as a new constructor of ``Type``::
+
+        | WitnessTy CoreExpr
+          -- typing rule:
+          -- Γ ⊢ e : exists (a :: ki). inner_ty
+          -- --------------------
+          -- Γ ⊢ WitnessTy e : ki
+
+      Note that this embeds expressions in types.
+
+   #. While packing and opening existentials is implicit in Haskell, it is explicit in Core,
+      using these two new constructors of ``Expr b``::
+
+        | Pack Type (Expr b) TyVar Type
+          -- typing rule:
+          -- Γ ⊢ witness_ty : ki
+          -- Γ ⊢ exists (bound_tv :: ki). inner_ty : TYPE rep
+          -- Γ ⊢ expr : inner_ty[witness_ty/bound_tv]
+          -- -----------------------------------------------------------------------
+          -- Γ ⊢ Pack witness_ty expr (bound_tv:ki) inner_ty : exists bound_tv. inner_ty
+
+        | Open (Expr b)
+          -- typing rule:
+          -- Γ ⊢ expr : exists (a :: k). inner_ty
+          -- -------------------------------------------------------------------------
+          -- Γ ⊢ Open expr : inner_ty[Witness @k (exists (a :: k). inner_ty) expr / a]
```

Comment:
Does this rule mean that core transforms (deep in `expr`) that previously preserved semantics and typing can now possibly change the type of `expr`? It seems as if this could significantly restrict GHC's leeway in optimizing expressions that _might_ occur in an existential type. Or am I missing something?

(I asked that at ICFP, but I don't quite remember the conclusion.)

Reply:
No, you're right. Perhaps this deserves more mention in the proposal. It is why `TransformEC` is so impossibly general, to allow arbitrary code transformations to be type-preserving.

Note that just about any further step toward dependent types would run into the same problem, so asserting the correctness of compiler transformations in order to preserve typing is in our future, regardless.

Reply:
Oh, it's right there, just further down. Should have read more carefully